pkg/crawler/crawler.go

```diff
@@ -262,3 +275,36 @@ func (c *Crawler) fetchSHA1(url string) ([]byte, error) {
 	}
 	return sha1b, nil
 }
+
+func (c *Crawler) fetchPOMLicense(url string) (string, error) {
+	resp, err := c.http.Get(url)
+	if resp.StatusCode == http.StatusNotFound {
+		return "", nil
+	}
+	if err != nil {
+		return "", xerrors.Errorf("can't get pom xml from %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+
+	var pomProject gopom.Project
+
+	decoder := xml.NewDecoder(resp.Body)
+	decoder.CharsetReader = charset.NewReaderLabel
+	err = decoder.Decode(&pomProject)
+
+	if err != nil {
+		return "", xerrors.Errorf("can't parse pom xml from %s: %w", url, err)
+	}
+
+	if len(pomProject.Licenses) == 0 {
+		return "", nil
+	}
+
+	var licenses []string
+	for _, l := range pomProject.Licenses {
+		licenses = append(licenses, l.Name)
+	}
+
+	// TODO: Check if we can limit the length of license string i.e trim and save
+	return strings.Join(licenses, ","), nil
```

Comment:
Looks like we need to think about this, because some licenses (e.g. `The Apache License, Version 2.0`) can contain commas. In this case will be confusion  if we will split by commas (e.g. `The Apache License, Version 2.0,The 2-Clause BSD License`).

Reply:
Right. The problem is that  there isn't a standard defined and the license section of pom.xml can contain effectively anything. 
Other option is to use urls instead of names. We can parse the urls to extract license information. But the problem is url might not be always available in  pom.xml .

We can probably have a preference order:
1. Pick url and parse it to extract license
2. If url is missing or not parsable , then pick license name and run it against the license classifier
3. If none of the conditions meet , then dump the license as is.

Let me know if you have any suggestions.

Reply:
We might have to call it out as a limitation I guess i.e we'll show whatever is captured in pom.xml. If you specify invalid attributes then that is what will be translated in the scan report.

As a precautionary step, we can add deduplication logic + limit number of licenses we will parse per GAV + a comma separated license string length check.

That might be less complicated in my opinion and easier to understand/comprehend.

Reply:
Gave license classifier a try.

It expects a license file . Does not result in matches using just the license name in POM
We essentially need to extract license URL from POM, fetch the contents of URL, dump it in a file and then use license classifier for the right results.

It will add an additional overhead. In some cases upto 2-3s per license. We might have to add an in memory  cache on top of it.

Reply:
> Integrating this in the current design of java db might be a problem, since concurrent goroutines for license classification might not work. It instantiates a backend instance , expects a list of license files, spawns go routines to process them and accumulates results in the backend.

This is really problem. We can add new step before crawler: parse all `pom.xml` files (as with `sha1` files) as you said (`Effectively a 3 step process:`) and save to map/file (url -> license name). And use this map/file in crawler. 
But there may be problem with CI/CD (work time, memory, etc...).

> The other perspective could be not over complicate things and claim support for whats highlighted in pom.xml. Since the user controls the pom.xml , it's their responsibility to specify the license information accurately. We will show whatever resides in pom.xml (license.Name) with some precautionary checks.

I think it is good solution. At least we can start with this
Maven docs say(https://maven.apache.org/pom.html#Licenses): `Using an SPDX identifier as the license name is recommended.`
It is not required, but we can refer on it.

Reply:
> This is really problem. We can add new step before crawler: parse all pom.xml files (as with sha1 files) as you said (Effectively a 3 step process:) and save to map/file (url -> license name). And use this map/file in crawler.
 But there may be problem with CI/CD (work time, memory, etc...).

Let me give this a try. We can probably crawl maven to find **unique** license URLs/Names across POM . Once all of them are fetched, we can use the license classifier for them and save the results in memory.

Same can then be used while crawling SHA1.

Reply:
Updated the PR.

1. Crawl to fetch GAV information + parse POM to extract license keys (hash of url if available , else name)
2. Save license keys along with GAV information in the indexes cache directory and simultaneously build a map of unique license keys with meta information about license name and url
3. Use license classifier to find license info associated with unique license keys found in step 2 and save them referenced by license key in a licenses cache directory
4. As part of build, parse the cached data in indexes directory. Extract license keys from the same and use it to extract license metadata from the licenses cache directory.
5. Write aggregated data to DB

Reply:
The DB size now is just 50MB more with license information(800MB) vs 750MB without license information.

| separated list of unique licenses : 
[licenses.csv](https://github.com/aquasecurity/trivy-java-db/files/12308455/licenses.csv)

The only problem seems to be time taken for building java db. https://github.com/aquasecurity/trivy-java-db/pull/21/files#diff-526a7142f7b15a72eab215dcfa540f5b6d502951a2f9f3b34dd900013e82c675R426 seems to be a major contributor other than additional POM parsing.

Don't think it should be a problem though since we refresh the database once a week.

Reply:
I think this is good way for us.